src/main/java/net/neoforged/neoforge/fluids/crafting/DataComponentFluidIngredient.java

```diff
@@ -0,0 +1,181 @@
+import java.util.Objects;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+import net.minecraft.core.Holder;
+import net.minecraft.core.HolderSet;
+import net.minecraft.core.component.DataComponentMap;
+import net.minecraft.core.component.DataComponentPredicate;
+import net.minecraft.core.component.DataComponentType;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.core.registries.Registries;
+import net.minecraft.resources.HolderSetCodec;
+import net.minecraft.world.level.material.Fluid;
+import net.neoforged.neoforge.common.NeoForgeMod;
+import net.neoforged.neoforge.common.crafting.DataComponentIngredient;
+import net.neoforged.neoforge.fluids.FluidStack;
+
+/**
+ * Fluid ingredient that matches the given set of fluids, additionally performing either a
+ * {@link DataComponentFluidIngredient#isStrict() strict} or partial test on the FluidStack's components.
+ * <p>
+ * Strict ingredients will only match fluid stacks that have <b>exactly</b> the provided components, while partial ones will
+ * match if the stack's components contain all required components for the {@linkplain #components input predicate}.
+ * 
+ * @see DataComponentIngredient DataComponentIngredient, its item equivalent
+ */
+public class DataComponentFluidIngredient extends FluidIngredient {
+    public static final MapCodec<DataComponentFluidIngredient> CODEC = RecordCodecBuilder.mapCodec(
+            builder -> builder
+                    .group(
+                            HolderSetCodec.create(Registries.FLUID, BuiltInRegistries.FLUID.holderByNameCodec(), false).fieldOf("fluids").forGetter(DataComponentFluidIngredient::fluids),
+                            DataComponentPredicate.CODEC.fieldOf("components").forGetter(DataComponentFluidIngredient::components),
+                            Codec.BOOL.optionalFieldOf("strict", false).forGetter(DataComponentFluidIngredient::isStrict))
+                    .apply(builder, DataComponentFluidIngredient::new));
+
+    private final HolderSet<Fluid> fluids;
+    private final DataComponentPredicate components;
+    private final boolean strict;
+    private final FluidStack[] stacks;
+
+    public DataComponentFluidIngredient(HolderSet<Fluid> fluids, DataComponentPredicate components, boolean strict) {
+        this.fluids = fluids;
+        this.components = components;
+        this.strict = strict;
+        this.stacks = fluids.stream()
+                .map(i -> new FluidStack(i, 1, components.asPatch()))
+                .toArray(FluidStack[]::new);
+    }
+
+    @Override
+    public boolean test(FluidStack stack) {
+        if (strict) {
+            for (FluidStack stack2 : this.stacks) {
+                if (FluidStack.isSameFluidSameComponents(stack, stack2)) return true;
+            }
+            return false;
+        } else {
+            return this.fluids.contains(stack.getFluidHolder()) && this.components.test(stack);
+        }
+    }
```

Comment:
Good point, that way we could ditch the cached stacks array as well!

Reply:
The problem here was that the patch can be different depending on which fluid is being compared.

(Well no because fluid stacks do not have a prototype at the moment; but if we do introduce one this will be a nasty bug).

I wouldn't worry too much about the `strict` path anyway, IMO non-strict is almost always a better choice.

Reply:
But what about the symmetry with item stacks? 😭

Reply:
Reverted this for now, it seems more future proof that way

Reply:
You can fix it for normal ingredients too if you want.

Reply:
Otherwise I'll make a quick PR; maybe that's better.